plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/precompiled/PrecompiledScriptPlugins.kt

```diff
@@ -28,51 +32,168 @@ import org.gradle.kotlin.dsl.precompile.PrecompiledSettingsScript
 import org.gradle.kotlin.dsl.support.ImplicitImports
 import org.gradle.kotlin.dsl.support.serviceOf
 
+import org.gradle.plugin.devel.GradlePluginDevelopmentExtension
+import org.gradle.plugin.devel.plugins.JavaGradlePluginPlugin
+
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
 import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 
+import java.io.File
+
 
 /*
- * Enables the compilation of `*.gradle.kts` scripts in regular Kotlin source-sets.
+ * Exposes `*.gradle.kts` scripts from regular Kotlin source-sets as binary Gradle plugins.
  */
 open class PrecompiledScriptPlugins : Plugin<Project> {
 
-    override fun apply(project: Project) = project.run {
+    override fun apply(project: Project): Unit = project.run {
 
-        afterEvaluate {
+        enableScriptCompilation()
 
-            tasks {
+        plugins.withType<JavaGradlePluginPlugin> {
+            exposeScriptsAsGradlePlugins()
+        }
+    }
+}
 
-                "compileKotlin"(KotlinCompile::class) {
-                    kotlinOptions {
-                        freeCompilerArgs += listOf(
-                            "-script-templates", scriptTemplates,
-                            // Propagate implicit imports and other settings
-                            "-Xscript-resolver-environment=${resolverEnvironment()}"
-                        )
-                    }
+
+private
+fun Project.enableScriptCompilation() {
+
+    afterEvaluate {
+
+        tasks {
+
+            "compileKotlin"(KotlinCompile::class) {
+                kotlinOptions {
+                    freeCompilerArgs += listOf(
+                        "-script-templates", scriptTemplates,
+                        // Propagate implicit imports and other settings
+                        "-Xscript-resolver-environment=${resolverEnvironment()}"
+                    )
                 }
             }
         }
     }
+}
+
+
+private
+val scriptTemplates by lazy {
+    listOf(
+        // treat *.settings.gradle.kts files as Settings scripts
+        PrecompiledSettingsScript::class.qualifiedName!!,
+        // treat *.init.gradle.kts files as Gradle scripts
+        PrecompiledInitScript::class.qualifiedName!!,
+        // treat *.gradle.kts files as Project scripts
+        PrecompiledProjectScript::class.qualifiedName!!
+    ).joinToString(separator = ",")
+}
+
+
+private
+fun Project.resolverEnvironment() =
+    (PrecompiledScriptDependenciesResolver.EnvironmentProperties.kotlinDslImplicitImports
+        + "=\"" + implicitImports().joinToString(separator = ":") + "\"")
+
+
+private
+fun Project.implicitImports(): List<String> =
+    serviceOf<ImplicitImports>().list
+
+
+private
+fun Project.exposeScriptsAsGradlePlugins() {
+
+    // TODO: script plugins from custom source-set dirs
```

Comment:
I'd be fine with handling a single source set for now.
In that case consider using `gradlePlugin.pluginSourceSet`.

Reply:
Makes a lot of sense. Done.